sky/shell/diagnostic/diagnostic_server.cc

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/shell/diagnostic/diagnostic_server.h"
+
+#include "base/bind.h"
+#include "base/location.h"
+#include "base/logging.h"
+#include "dart/runtime/include/dart_api.h"
+#include "dart/runtime/include/dart_native_api.h"
+#include "flow/paint_context.h"
+#include "sky/engine/core/script/embedder_resources.h"
+#include "sky/engine/tonic/dart_binding_macros.h"
+#include "sky/engine/tonic/dart_invoke.h"
+#include "sky/engine/tonic/dart_library_natives.h"
+#include "sky/shell/rasterizer.h"
+#include "sky/shell/shell.h"
+#include "sky/shell/gpu/picture_serializer.h"
+#include "sky/shell/ui/engine.h"
+#include "third_party/skia/include/core/SkStream.h"
+
+namespace mojo {
+namespace dart {
+extern ResourcesEntry __sky_embedder_diagnostic_server_resources_[];
+}
+}
+
+namespace sky {
+namespace shell {
+
+using blink::DartLibraryNatives;
+using blink::EmbedderResources;
+using blink::DartInvokeField;
+using blink::LogIfError;
+
+namespace {
+
+DartLibraryNatives* g_natives = nullptr;
+
+const char kDiagnosticServerScript[] = "/diagnostic_server.dart";
+
+EmbedderResources resources(
+    &mojo::dart::__sky_embedder_diagnostic_server_resources_[0]);
+
+Dart_NativeFunction GetNativeFunction(Dart_Handle name,
+                                      int argument_count,
+                                      bool* auto_setup_scope) {
+  CHECK(g_natives);
+  return g_natives->GetNativeFunction(name, argument_count, auto_setup_scope);
+}
+
+const uint8_t* GetSymbol(Dart_NativeFunction native_function) {
+  CHECK(g_natives);
+  return g_natives->GetSymbol(native_function);
+}
+
+void SendNull(Dart_Port port_id) {
+  Dart_CObject null_object;
+  null_object.type = Dart_CObject_kNull;
+  Dart_PostCObject(port_id, &null_object);
+}
+
+}  // namespace
+
+DART_NATIVE_CALLBACK_STATIC(DiagnosticServer, HandleSkiaPictureRequest);
+
+void DiagnosticServer::Start() {
```

Comment:
So, if I understand this correctly, we would package in the script that implements the diagnostics server and create a library for the same at runtime?

This would not work in the AOT runtime however. Can we package this script like we handle sky_engine and sky_service? That way, the precompiler can pick them up.